src/Intl/Idn/Idn.php

```diff
@@ -0,0 +1,316 @@
+
+/**
+ * Partial intl implementation in pure PHP.
+ *
+ * Implemented:
+ * - idn_to_ascii - Convert domain name to IDNA ASCII form
+ * - idn_to_utf8  - Convert domain name from IDNA ASCII to Unicode
+ *
+ * @author Renan Gonçalves <renan.saddam@gmail.com>
+ * @author Sebastian Kroczek <sk@xbug.de>
+ * @author Dmitry Lukashin <dmitry@lukashin.ru>
+ * @author Laurent Bassin <laurent@bassin.info>
+ *
+ * @internal
+ */
+final class Idn
+{
+    const BASE = 36;
+    const TMIN = 1;
+    const TMAX = 26;
+    const SKEW = 38;
+    const DAMP = 700;
+    const INITIAL_BIAS = 72;
+    const INITIAL_N = 128;
+    const PREFIX = 'xn--';
+    const DELIMITER = '-';
+
+    private static $encodeTable = array(
+        'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l',
+        'm', 'n', 'o', 'p', 'q', 'r', 's', 't', 'u', 'v', 'w', 'x',
+        'y', 'z', '0', '1', '2', '3', '4', '5', '6', '7', '8', '9',
+    );
+
+    private static $decodeTable = array(
+        'a' => 0, 'b' => 1, 'c' => 2, 'd' => 3, 'e' => 4, 'f' => 5,
+        'g' => 6, 'h' => 7, 'i' => 8, 'j' => 9, 'k' => 10, 'l' => 11,
+        'm' => 12, 'n' => 13, 'o' => 14, 'p' => 15, 'q' => 16, 'r' => 17,
+        's' => 18, 't' => 19, 'u' => 20, 'v' => 21, 'w' => 22, 'x' => 23,
+        'y' => 24, 'z' => 25, '0' => 26, '1' => 27, '2' => 28, '3' => 29,
+        '4' => 30, '5' => 31, '6' => 32, '7' => 33, '8' => 34, '9' => 35
+    );
+
+    public static function idn_to_ascii($domain, $options, $variant, &$idna_info = array())
+    {
+        $parts = explode('.', $domain);
+
+        foreach ($parts as &$part) {
+            $length = strlen($part);
+            if ($length < 1) {
+                return false;
+            }
+            $part = self::encodePart($part);
+            if ($part === false) {
+                return false;
+            }
+        }
+
+        $output = implode('.', $parts);
+        $length = strlen($output);
+        if ($length > 255) {
+            return false;
+        }
+
+        return $output;
+    }
+
+    public static function idn_to_utf8($domain, $options, $variant, &$idna_info = array())
+    {
+        if ($variant !== INTL_IDNA_VARIANT_2003) {
+            return false;
+        }
+
+        $parts = explode('.', $domain);
+        foreach ($parts as &$part) {
+            $length = strlen($part);
+            if ($length > 63 || $length < 1) {
+                continue;
+            }
+            if (strpos($part, static::PREFIX) !== 0) {
+                continue;
+            }
+
+            $part = substr($part, strlen(static::PREFIX));
+            $part = self::decodePart($part);
+        }
+        $output = implode('.', $parts);
+        $length = strlen($output);
+        if ($length > 255) {
+            return false;
+        }
+
+        return $output;
+    }
+
+    private static function encodePart($input)
+    {
+        $codePoints = self::listCodePoints($input);
+
+        $n = static::INITIAL_N;
```

Comment:
should use `self::`. It does not make sense to support overriding the constants, so using late static binding is useless (and removes opportunities for OPCache to optimize things as it forces runtime evaluation)